services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_volume_removal.py

```diff
@@ -0,0 +1,158 @@
+
+        volume_removal_service = await client.services.create(
+            **jsonable_encoder(service_spec, by_alias=True, exclude_unset=True)
+        )
+
+        service_id = volume_removal_service["ID"]
+        try:
+            async for attempt in AsyncRetrying(
+                stop=stop_after_delay(service_timeout_s),
+                wait=wait_fixed(0.5),
+                retry=retry_if_exception_type(_RetryError),
+                reraise=True,
+            ):
+                with attempt:
+                    tasks = await client.tasks.list(filters={"service": service_id})
+                    # NOTE: the service will have at most 1 task, since there is no restart
+                    # policy present
+                    if len(tasks) != 1:
+                        # Sometimes swarm did not mange to create a task after the service
+                        # is created. In that case a retry is triggered
+                        raise _RetryError(
+                            f"Expected 1 task for service {service_id}, found {tasks=}"
+                        )
+
+                    task = tasks[0]
+                    task_status = task["Status"]
+                    log.debug("Service %s, %s", service_id, f"{task_status=}")
+                    task_state = task_status["State"]
+                    if task_state not in SERVICE_FINISHED_STATES:
+                        raise _RetryError(f"Waiting for task to finish: {task_status=}")
+
+                    if not (
+                        task_state == "complete"
+                        and task_status["ContainerStatus"]["ExitCode"] == 0
+                    ):
+                        # recover logs from command for a simpler debugging
+                        container_id = task_status["ContainerStatus"]["ContainerID"]
+                        container = await client.containers.get(container_id)
+                        container_logs = await container.log(stdout=True, stderr=True)
+                        log.error(
+                            "Service %s, %s output: %s",
+                            service_id,
+                            f"{task_status=}",
+                            "\n".join(container_logs),
+                        )
+                        # NOTE: above implies the volumes will remain in the system and
+                        # have to be manually removed.
+                        return False
+        finally:
+            # NOTE: services created in swarm need to be removed, there is no way
+            # to instruct swarm to remove a service after it's created
+            # container/task finished
+            await client.services.delete(service_id)
+
+        return True
+
+
+async def remove_pending_volume_removal_services(
+    dynamic_sidecar_settings: DynamicSidecarSettings,
+) -> None:
+    """
+    Removes all pending volume removal services. Such a service
+    will be considered pending if it is running for longer than its
```

Comment:
so if I get it right, that would be if the fct above that one failed? or if the director-v2 is restarted?

Reply:
That is correct.